### PR TITLE
Ensure function parameter names always matches parameter types length

### PR DESCRIPTION
**Pull request by Marenz**

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1234,10 +1234,10 @@ bool TypeChecker::visit(TupleExpression const& _tuple)
 					if (!dynamic_cast<RationalNumberType const&>(*types[i]).mobileType())
 						m_errorReporter.fatalTypeError(components[i]->location(), "Invalid rational number.");
 
-				if (_tuple.isInlineArray())
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:1237]: Haha, nice one! :)

File: libsolidity/ast/Types.h [hunk 2]
```diff
@@ -1050,6 +1050,13 @@ class FunctionType: public Type
 		m_bound(_bound),
 		m_declaration(_declaration)
 	{
+		if (m_parameterNames.empty() && !m_parameterTypes.empty())
+			m_parameterNames = strings(m_parameterTypes.size(), "");
```
Comment (chriseth) [inline libsolidity/ast/Types.h]: I would prefer to do this filling in the special-purpose constructor above only ("`Function type constructor to be used for a plain type (not derived from a declaration).`") and add the assertion here.

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -1678,17 +1678,10 @@ void TypeChecker::typeCheckFunctionGeneralChecks(
...
-		{
-			m_errorReporter.typeError(
-				_functionCall.location(),
-				parameterNames.size() > argumentNames.size() ?
-				"Some argument names are missing." :
-				"Too many arguments."
-			);
-			return;
-		}
+		solAssert(
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:1681]: Where is the check that allows you to turn this into an assertion?

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/analysis/TypeChecker.cpp (hunk 3 above)
Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp:1681]: If we can assume that the parameter names list is the same lengths as the parameter types list, which we can with the newly added asserts in the c'tor, then line 1593 does this check already: https://github.com/ethereum/solidity/pull/5957/files#diff-9f2a45efaa7b75ab7ec620d3177cf01aL1593

File: libsolidity/ast/Types.h (hunk 2 above)
Comment (Marenz) [reply libsolidity/ast/Types.h]: I am now working on that. So far it requires fixes all over the code because both c'tors are used

**Comment by Marenz** [comment]
Updated:
* Added the same assert to all "final" c'tors
* Fixed code all over to make sure the lengths are correct
* Also added the fix for the *return* parameter name list

**Comment by Marenz** [comment]
Update: Hopefully fixed the CI failure

**Review by chriseth — APPROVED** [review]

**Pull request merged**